### PR TITLE
Add field constraints

### LINKED ISSUE #135: [PE-D][Tester C] Phone number can be all 0s
I am able to add Persons whose number is all 0. For example 000 is accepted as a valid Phone Number as shown in the attached picture. I think this should be a relatively simple fix by editing the validation regex!

Image

[TESTED ON MAC]

-------------
Labels: `severity.Low` `type.FunctionalityBug`

### LINKED ISSUE #162: [PE-D][Tester B] Contacts with common names containing hyphens or apostrophes cannot be added
Environment:
- App version: v1.5.1
- OS: Windows 11
- Java: 17

Steps to reproduce:
1. Launch the app.
2. Enter `:add n/Anne-Marie O'Neil p/91234567 e/anne@example.com a/NUS Computing`.

Expected:
The contact should be added successfully, since hyphens and apostrophes are common in real names.

Actual:
The command is rejected because the name is considered invalid.

Why this is a bug:
This is overzealous input validation for a contact-management app. It prevents users from storing legitimate names and makes the feature less useful in normal usage.

-------------
Labels: `severity.Medium` `type.FeatureFlaw`

### LINKED ISSUE #183: [PE-D][Tester B] Contacts with common international phone formats cannot be added
Environment:
- App version: v1.5.1
- OS: Windows 11
- Java: 17

Steps to reproduce:
1. Enter `:add n/John Doe p/+65 91234567 e/john@example.com a/NUS Computing`.

Expected:
The contact should be added successfully, since `+65 91234567` is a common real-world phone format.

Actual:
The command is rejected because the phone number is considered invalid.

Why this is a bug:
This is overzealous input validation for a contact-management app. Users commonly store numbers with country codes and spacing, so rejecting them makes the app less useful in normal usage.

-------------
Labels: `severity.Medium` `type.FeatureFlaw`

### LINKED ISSUE #184: [PE-D][Tester B] Contacts with non-English names cannot be added
Environment:
- App version: v1.5.1
- OS: Windows 11
- Java: 17

Steps to reproduce:
1. Enter `:add n/José Álvarez p/91234567 e/jose@example.com a/NUS Computing`.

Expected:
The contact should be added successfully, since names with accented characters are legitimate real-world names.

Actual:
The command is rejected because the name is considered invalid.

Why this is a bug:
This is overzealous input validation for a contact-management app. It prevents users from storing legitimate names and makes the app less useful in normal usage.

-------------
Labels: `severity.Medium` `type.FeatureFlaw`

### LINKED ISSUE #195: [PE-D][Tester C] Long Names cannot be fully displayed on the UI
Sufficiently long names cannot be seen on the left hand side of the UI (however, it can be seen in the details section) as shown in the image below.

Image

Another potential fix might just to simply allow users to scroll horizontally for the left side of the UI if needed. Not too sure on the need for this complexity too as names should not be so terribly long 

-------------
Labels: `severity.Medium` `type.FeatureFlaw`

### LINKED ISSUE #196: [PE-D][Tester C] Phone numbers have no limit length
The phone numbers can be of any length, which can warp the UI as well as seen in the image below

Image

I edited the number to be extremely long and now the label title itself cannot be seen as well. 

A easy fix would definitely be to cap the length to a reasonable length (maybe 50?)!

-------------
Labels: `severity.Medium` `type.FeatureFlaw`

### LINKED ISSUE #198: [PE-D][Tester C] Long emails are allowed
Similar to the other fields, long emails are allowed to be input by the User and it can warp the UI (and not being fully displayed as well) as seen in the image below

Image

-------------
Labels: `severity.Medium` `type.FeatureFlaw`

### PR DESCRIPTION
**Pull request by eunicegwee**
### Summary
This PR adds stricter constraints to the name, phone number, email and address fields.

#### Name
Constraints:

- Names must not be blank.
- Names can contain Unicode letters and digits (supports names from different languages/scripts).
- Allowed punctuation in names: spaces, hyphens (-), apostrophes ('), and periods (.).
- Other special characters are not allowed.
- Names must start with an alphanumeric character.
- Names have a character limit of 100.

Closes: #195 
Closes: #184 
Closes: #162 

#### Phone
Constraints:

- Phone numbers must use ASCII digits 0-9 only.
- Phone numbers must be 3 to 15 digits long.
- An optional leading + is allowed, but only as the first character.
- If + is present, the next character must be a digit from 1 to 9.
- Phone numbers cannot be all zeros.
- No spaces, letters, punctuation, or other symbols are allowed.

Closes: #183 
Closes: #135 
Closes: #196 

#### Email

- Must be in the form local-part@domain.
- Local-part may contain only ASCII letters, digits, and the special characters +, _, ., -.
- Local-part must not start/end with a special character, and special characters cannot appear consecutively.
- Local-part must be at most 64 characters long.
- Domain must contain at least one . so it has both a second-level domain and a top-level domain.
- Domain may contain only ASCII letters, digits, hyphens, and periods.
- Each domain label must start and end with an alphanumeric character; hyphens may appear only inside a label.
- The top-level domain must be at least 2 characters long.
- Domain must be at most 253 characters long.
- Entire email address must be at most 254 characters long.

Closes: #198 

#### Address

- Address must be 1 to 255 characters long.
- Address must contain at least one alphanumeric character.
- Address cannot contain any ISO control characters.
- Spaces and punctuation are allowed as long as the above rules are satisfied.

### CODE REVIEW

File: src/main/java/seedu/address/model/person/Phone.java [hunk 1]
```diff
@@ -11,8 +11,8 @@ public class Phone {
 
 
     public static final String MESSAGE_CONSTRAINTS =
-            "Phone numbers should only contain numbers, and it should be at least 3 digits long";
-    public static final String VALIDATION_REGEX = "\\d{3,}";
+            "Phone numbers should be 3 to 15 digits, must not be all zeros, and may start with '+' followed by 1-9.";
+    public static final String VALIDATION_REGEX = "(?:\\+[1-9][0-9]{2,14}|(?!0{3,15}$)[0-9]{3,15})";
```
Comment (Jaepple) [inline src/main/java/seedu/address/model/person/Phone.java:15]: Great use of Regex here!

File: src/main/java/seedu/address/model/person/Name.java [hunk 2]
```diff
@@ -10,13 +10,16 @@
...
+            "Names should start with a letter or number, contain only letters, numbers, spaces, hyphens (-), "
+                    + "apostrophes ('), and periods (.), and be at most 100 characters long";
+
+    public static final int MAX_LENGTH = 100;
 
     /*
      * The first character of the address must not be a whitespace,
      * otherwise " " (a blank string) becomes a valid input.
      */
-    public static final String VALIDATION_REGEX = "[\\p{Alnum}][\\p{Alnum} ]*";
```
Comment (kavishsathia) [inline src/main/java/seedu/address/model/person/Name.java:19]: Users might paste a different quote (' or ')

File: src/main/java/seedu/address/model/person/Address.java [hunk 3]
```diff
@@ -9,13 +9,15 @@
...
+    public static final String MESSAGE_CONSTRAINTS =
+            "Addresses should be at most 255 characters, cannot contain ISO control characters, "
+                    + "and must contain at least one alphanumeric character";
 
     /*
      * The first character of the address must not be a whitespace,
      * otherwise " " (a blank string) becomes a valid input.
      */
-    public static final String VALIDATION_REGEX = "[^\\s].*";
+    public static final String VALIDATION_REGEX = "(?=.{1,255}$)(?=.*[\\p{L}\\p{N}])(?!.*\\p{Cntrl})[^\\s].*";
```
Comment (Jaepple) [inline src/main/java/seedu/address/model/person/Address.java:20]: Similarly great use of regex to easily enforce the constraints

**Review by Jaepple — COMMENTED** [review]
LGTM! Merge!

File: src/main/java/seedu/address/model/person/Email.java [hunk 4]
```diff
@@ -48,7 +47,26 @@ public Email(String email) {
      * Returns if a given string is a valid email.
      */
     public static boolean isValidEmail(String test) {
```
Comment (kavishsathia) [inline src/main/java/seedu/address/model/person/Email.java:49]: test@localhost is not a valid email anymore, I think we can add an issue to update this in the UG because this is an edge case anyways. We should only allow emails with domain names that are publicly accessible and have a TLD.
